test/e2e/ownerreference_test.go

```diff
@@ -0,0 +1,282 @@
+	})
+
+})
+
+var (
+	VSphereKubernetesReferenceAssertions = map[string]func([]metav1.OwnerReference) error{
+		// Need custom Kubernetes assertions for secrets. Secrets in the CAPV tests can also be owned by the vSphereCluster.
+		"Secret": func(owners []metav1.OwnerReference) error {
+			return framework.HasOneOfExactOwners(owners,
+				// Secrets for cluster certificates must be owned by the KubeadmControlPlane.
+				[]metav1.OwnerReference{kubeadmControlPlaneController},
+				// The bootstrap secret should be owned by a KubeadmConfig.
+				[]metav1.OwnerReference{kubeadmConfigController},
+				// Secrets created as a resource for a ClusterResourceSet can be owned by the ClusterResourceSet.
+				[]metav1.OwnerReference{clusterResourceSetOwner},
+				// Secrets created as an identityReference for a vSphereCluster should be owned but the vSphereCluster.
+				[]metav1.OwnerReference{vSphereClusterOwner},
+			)
+		},
+		"ConfigMap": func(owners []metav1.OwnerReference) error {
+			// The only configMaps considered here are those owned by a ClusterResourceSet.
+			return framework.HasExactOwners(owners, clusterResourceSetOwner)
+		},
+	}
+)
+
+var (
+	VSphereReferenceAssertions = map[string]func([]metav1.OwnerReference) error{
+		"VSphereCluster": func(owners []metav1.OwnerReference) error {
+			return framework.HasExactOwners(owners, clusterController)
+		},
+		"VSphereClusterTemplate": func(owners []metav1.OwnerReference) error {
+			return framework.HasExactOwners(owners, clusterClassOwner)
+		},
+		"VSphereMachine": func(owners []metav1.OwnerReference) error {
+			// The vSphereCluster takes ownership of all vSphereMachines in addition to the core Machine.
+			return framework.HasExactOwners(owners, vSphereClusterOwner, machineController)
+		},
+		"VSphereMachineTemplate": func(owners []metav1.OwnerReference) error {
+			// The vSphereMachineTemplate can be owned by the Cluster or the ClusterClass.
+			return framework.HasOneOfExactOwners(owners, []metav1.OwnerReference{clusterOwner}, []metav1.OwnerReference{clusterClassOwner})
+		},
+		"VSphereVM": func(owners []metav1.OwnerReference) error {
+			return framework.HasExactOwners(owners, vSphereMachineOwner)
+		},
+		// VSphereClusterIdentity does not have any owners.
+		"VSphereClusterIdentity": func(owners []metav1.OwnerReference) error {
+			// The vSphereClusterIdentity does not have any owners.
+			return framework.HasExactOwners(owners)
+		},
+		"VSphereDeploymentZone": func(owners []metav1.OwnerReference) error {
+			// The vSphereDeploymentZone does not have any owners.
+			return framework.HasExactOwners(owners)
+		},
+		"VSphereFailureDomain": func(owners []metav1.OwnerReference) error {
+			// The vSphereFailureDomain can be owned by one or more vSphereDeploymentZones.
+			return framework.HasOneOfExactOwners(owners, []metav1.OwnerReference{vSphereDeploymentZoneOwner}, []metav1.OwnerReference{vSphereDeploymentZoneOwner, vSphereDeploymentZoneOwner})
```

Comment:
nit: Up to you but in our case it's only one. So I think just asserting on one would be better/more precise

Reply:
More precise but less useful. I think it's a good idea to document, at least in a limited way, the possible ownership chains in these tests, but fine to remove it if you feel strongly.

Reply:
supporting 1 & 2 owners just seems like an arbitrary number

Reply:
But it's fine